modules/distributor/usage/tracker.go

```diff
@@ -0,0 +1,425 @@
+	labels []string
+
+	// Runtime data
+	bytes       uint64
+	lastUpdated int64
+}
+
+func (b *bucket) Inc(bytes uint64, unix int64) {
+	b.bytes += bytes
+	b.lastUpdated = unix
+}
+
+type mapping struct {
+	from string
+	to   int // Index into the values buffer
+}
+
+type tenantUsage struct {
+	series      map[uint64]*bucket
+	constLabels prometheus.Labels
+
+	// Buffers for Observe
+	dimensions map[string]string // Originally configured dimensions
+	mapping    []mapping         // Mapping from attribute => final sanitized label. Typically few values and slice is faster than map
+	sortedKeys []string          // So we can always iterate the buffer in order, this can be precomputed up front
+	buffer1    []string          // Batch-level values
+	buffer2    []string          // Span-level values
+	buffer3    []string          // Last hashed values
+	overflow   uint64
+}
+
+// GetBuffersForDimensions takes advantage of the fact that the configuration for a tracker
+// changes slowly.  Reuses buffers from the previous call when the dimensions are the same.
+func (t *tenantUsage) GetBuffersForDimensions(dimensions map[string]string) ([]mapping, []string, []string, []string) {
+	if !maps.Equal(dimensions, t.dimensions) {
+		// The configuration changed.
+
+		// Step 1
+		// Gather all configured dimensions and their sanitized output
+		t.dimensions = dimensions
+		sanitizedDimensions := make(map[string]string, len(dimensions))
+		for k, v := range dimensions {
+			// Get the final sanitized output label for this
+			// dimension.  Dimensions are key-value pairs with
+			// optional value.  If value is empty string, then
+			// we use the just the key.  Regardless the output
+			// is always the sanitized version.
+			// Example:
+			//    service.name="" 			=> "service_name"
+			//    service.name="foo.bar"	=> "foo_bar"
+			var sanitized string
+			if v == "" {
+				// The dimension is using default mapping
+				v = k
+			}
+			sanitized = strutil.SanitizeFullLabelName(v)
+			sanitizedDimensions[k] = sanitized
+		}
+
+		// Step 2
+		// Build the final list of sorted/distinct outputs
+		t.sortedKeys = t.sortedKeys[:0]
+		for _, v := range sanitizedDimensions {
+			if !slices.Contains(t.sortedKeys, v) {
+				t.sortedKeys = append(t.sortedKeys, v)
+			}
+		}
+		slices.Sort(t.sortedKeys)
+
+		// Step 3
+		// Prepare the mapping from raw attribute names to the final location of
+		// where it goes in the output buffers. This avoids another layer of indirection.
+		t.mapping = t.mapping[:0]
+		for k := range dimensions {
+			i := slices.Index(t.sortedKeys, sanitizedDimensions[k])
+			t.mapping = append(t.mapping, mapping{
+				from: k,
+				to:   i,
+			})
+		}
+
+		// Step 4
+		// Prepopulate the buffers and precompute the overflow bucket
+		t.buffer1 = make([]string, len(t.sortedKeys))
+		t.buffer2 = make([]string, len(t.sortedKeys))
+		t.buffer3 = make([]string, len(t.sortedKeys))
```

Comment:
reuse existing buffer? i think i'm just getting mixed messages about allocs b/c we're reusing t.sortedKeys but none of the other fields.

Reply:
nevermind. i see it used below. i thought it was local only to this method.